### PR TITLE
Implement FoundationDB backing store

### PR DESCRIPTION
**Pull request by masih**
Implement FoundationDB backing store that uses dedicated prefix directories for multihashes and metadata. Adjust `Dockerfile` to include fdb client C bindings.

Refactor the codebase to separate packages.

### CODE REVIEW

File: .github/workflows/go-test-config.json [hunk 1]
```diff
@@ -1,3 +1,7 @@
 {
-  "skip32bit": true
+  "skip32bit": true,
+  "skipOSes": [
+    "windows",
+    "macos"
```
Comment (willscott) [inline .github/workflows/go-test-config.json:5]: is there a reason it doesn't work on mac?

File: cmd/dhstore/main.go [hunk 2]
```diff
@@ -28,71 +31,88 @@ func main() {
 	dwal := flag.Bool("disableWAL", false, "Weather to disable WAL in Pebble dhstore.")
 	blockCacheSize := flag.String("blockCacheSize", "1Gi", "Size of pebble block cache. Can be set in Mi or Gi.")
 	llvl := flag.String("logLevel", "info", "The logging level. Only applied if GOLOG_LOG_LEVEL environment variable is unset.")
+	useFDB := flag.Bool("useFDB", false, "Weather to use FoundationDB database. Embedded Pebble store used by default. If enabled, fdb* flags must be set.")
```
Comment (willscott) [inline cmd/dhstore/main.go:34]: this is probably better as a string of which db backend

**Review by willscott — COMMENTED** [review]
this looks good.

@ischasny  should review/approve

File: fdb/fdb.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+	}
+	_, err = f.db.Transact(func(transaction fdb.Transaction) (any, error) {
+		// Hash encrypted value key to fix the size of key tuple. It may be
+		// more efficient to store the encrypted value key as is if it's
+		// short enough. But hashing it would offer a more even utilisation
+		// of the cluster by eliminating potential common prefixes in the
+		// encrypted value key.
+		// TODO: investigate if there are benefits here to be explored in
+		//       terms of reduction in disk utilisation.
+		vkh, err := f.hash(vk)
```
[the remark this answers is not in the document]
Comment (ischasny) [reply fdb/fdb.go:78]: The original key should be short enough without second layer of hashing. We also know that it's aways 32 bytes (sha256)

File: fdb/fdb.go [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+func NewFDBDHStore(o ...Option) (*FDBDHStore, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	if err := fdb.APIVersion(opts.apiVersion); err != nil {
+		return nil, err
+	}
+	db, err := fdb.OpenDatabase(opts.clusterFile)
```
Comment (ischasny) [inline fdb/fdb.go:50]: Am I right in understanding that we are going to have to do an extra hop on the write path? I.e. the write request would hit `dhstore`, then `dhstore` would write into its local FDB and then the local FDB would route that write to the correct node in the consistent hashing ring? 

If so - do we know whether that impacts write throughput in any way? 

Same question with regards to the reads - is there a way to route a read request to the right node in the cluster straight up?

File: bench/bench_test.go [hunk 5]
```diff
@@ -159,29 +160,29 @@ func benchmarkGetMetadatas(b *testing.B, n, hvkLen, mdLen int) {
 	b.StopTimer()
 }
 
-func getMultihashes(b *testing.B, mhs []multihash.Multihash, store *dhstore.PebbleDHStore) {
+func getMultihashes(b *testing.B, mhs []multihash.Multihash, store dhstore.DHStore) {
```
Comment (ischasny) [inline bench/bench_test.go:163]: I wanted to run the `bench_test` with the FDB store, however it kept failing with missing C bindings
```
/Users/ivan/go/pkg/mod/github.com/apple/foundationdb/bindings/go@v0.0.0-20230508183941-1098aa0251f4/src/fdb/database.go:26:11: fatal error: 'foundationdb/fdb_c.h' file not found
 #include <foundationdb/fdb_c.h>
          ^~~~~~~~~~~~~~~~~~~~~~
```

Am I missing something in my local environment / is there a way to run it on Mac?

File: .github/workflows/go-test-config.json (hunk 1 above)
Comment (masih) [reply .github/workflows/go-test-config.json:5]: It does work on both mac and windows. It just needs installation of c bindings. I chose to postpone setting this up for CI in favour of testing this in dev to see if it fits the task in hand. _if_ it does then I intend to attend this debt.

File: .github/workflows/go-test.yml [hunk 6]
```diff
@@ -28,6 +28,14 @@ jobs:
...
+      - name: Install FoundationDB Client C bindings
+        env:
+          FDB_VERSION: '7.1.33'
+          FDB_CLIENTS_DEB_SHA256_SUM: 'd73002bc796de7ce0158f1eb64fab1f8e90bada2edd90599d9b40754220150af'
+        run: |
+          wget "https://github.com/apple/foundationdb/releases/download/${FDB_VERSION}/foundationdb-clients_${FDB_VERSION}-1_amd64.deb"
+          echo "${FDB_CLIENTS_DEB_SHA256_SUM}  foundationdb-clients_${FDB_VERSION}-1_amd64.deb" | sha256sum --check
+          sudo dpkg --force-all -i foundationdb-clients_${FDB_VERSION}-1_amd64.deb
       - name: Use msys2 on windows
         if: matrix.os == 'windows'
```
[the remark this answers is not in the document]
Comment (masih) [reply .github/workflows/go-test.yml:40]: I wanted to refrain from modifying autogenerated CI builds as much as I could. Piotr has kindly given me pointer on how to avoid this. 

If sorting out the CI build on mac and windows is a non blocker I'd like to push ahead with testing this on dev before spending more time on FDB.

File: fdb/fdb.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (masih) [reply fdb/fdb.go:78]: Sure. Do we have a max expected size for `metadata` key? I believe those are not necessarily multihashes. I would like to enforce max supported in fdb since it impacts performance. Otherwise, i'll keep the hashing for metadata keys. 

Thoughts?

File: fdb/fdb.go (hunk 4 above)
Comment (masih) [reply fdb/fdb.go:50]: There is no local FDB. I believe the client library does batching internally i.e. does not do a remote call per `Set`. 

I think it's far more effective to get this running in `dev` and let the numbers talk.

File: bench/bench_test.go (hunk 5 above)
Comment (masih) [reply bench/bench_test.go:163]: Yup, your setup is missing the c binding installation. Take a look at the changes I have made in the CI on how to install them. There are good docs on FDB website on installing binaries. 

You can also look at `Dockerfile` on how they are installed. Note that they are platform dependant. You will need arm64 for M1 if that's what you are running?

File: fdb/fdb.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (masih) [reply fdb/fdb.go:78]: Got it, but it's not a multihash right? 

Going with 32 bytes max length enforcement.

File: bench/bench_test.go (hunk 5 above)
Comment (ischasny) [reply bench/bench_test.go:163]: Ok - will have a look thanks

File: fdb/fdb.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ischasny) [reply fdb/fdb.go:78]: It's not a multihash, just fixed 32 byte array

File: bench/bench_test.go [hunk 7]
```diff
@@ -159,29 +160,29 @@ func benchmarkGetMetadatas(b *testing.B, n, hvkLen, mdLen int) {
...
 	for _, mh := range mhs {
 		evks, err := store.Lookup(mh)
 		require.NoError(b, err)
 		require.NotNil(b, evks)
 	}
 }
 
-func putMultihashes(b *testing.B, mhs []multihash.Multihash, vks [][]byte, store *dhstore.PebbleDHStore) {
+func putMultihashes(b *testing.B, mhs []multihash.Multihash, vks [][]byte, store dhstore.DHStore) {
```
Comment (gammazero) [inline bench/bench_test.go:171]: What is the reason for not passing in a `DHStore`  interface?

File: fdb/fdb.go [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+	if multicodec.Code(dmh.Code) != multicodec.DblSha2_256 {
+		return dhstore.ErrUnsupportedMulticodecCode{Code: multicodec.Code(dmh.Code)}
+	}
+	if dmh.Length != 32 {
+		return dhstore.ErrMultihashDecode{Err: errMultihashDigestLength, Mh: mh}
+	}
+	if len(vk) > fdbMaxValueBytes {
+		return fmt.Errorf("value key cannot be larger than 100 KB, got: %d", len(vk))
+	}
+	_, err = f.db.Transact(func(transaction fdb.Transaction) (any, error) {
```
Comment (gammazero) [inline fdb/fdb.go:79]: This is doing a separate transaction for each individual multihash. Seems like the transaction should be wrapped around a batch of multihashes. Seems like separate transactions would cause separate db connections, and defeat what we want to use FDB for in the first place.

File: fdb/fdb.go [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+		// TODO: implement streaming variation since we are dealing with a streaming iterator anyway.
+		iterator := vks.Iterator()
+		var evks []dhstore.EncryptedValueKey
+		var latestErr error
+		for iterator.Advance() {
+			kv, err := iterator.Get()
+			if err != nil {
+				latestErr = err
+				logger.Errorw("failed to list encrypted value keys for multihash", "mh", mh.B58String(), "err", err)
+			}
```
Comment (gammazero) [inline fdb/fdb.go:173]: Need `continue`?

File: fdb/fdb.go [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+					logger.Errorw("expected unpacked key of length 2 ", "len", len(unpack), "mh", mh.B58String())
+				} else {
+					v, ok := unpack[1].([]byte)
+					if !ok {
+						logger.Errorw("expected unpacked key type bytes ", "got", unpack[0], "mh", mh.B58String())
+					} else {
+						evks = append(evks, v)
+					}
+				}
+			} else {
```
Comment (gammazero) [inline fdb/fdb.go:191]: style: Use of `continue` instead of `else` would reduce indent.

File: fdb/fdb.go (hunk 8 above)
Comment (masih) [reply fdb/fdb.go:79]: I read on this a bit and have some ideas.
 I'd like to get some basic numbers out before spending more time. 
 
 TL;DR 
* the recommendation is to use concurrent goroutines each open a transaction.
* In our case repeated multihashes are relatively rare so writing the entire batch in one transaction seems reasonable.
* this needs interface changes across dhstore which i'd like to hold off since it causes conflicts with other open PRs.

File: bench/bench_test.go (hunk 7 above)
Comment (masih) [reply bench/bench_test.go:171]: not sure i follow; `dhstore.DHStore` is the interface.

File: Dockerfile [hunk 11]
```diff
@@ -1,14 +1,22 @@
 FROM golang:1.19-bullseye as build
```
[the remark this answers is not in the document]
Comment (masih) [reply Dockerfile:1]: Yes once unified CI upgrades. For now the version here matches gomod.

File: .github/workflows/go-test-config.json (hunk 1 above)
Comment (masih) [reply .github/workflows/go-test-config.json:5]: https://github.com/ipni/dhstore/issues/48

File: fdb/fdb.go (hunk 8 above)
Comment (masih) [reply fdb/fdb.go:79]: Also see:  [FDB Read/Write Path](https://apple.github.io/foundationdb/read-write-path.html?highlight=batch)

> Client. It is a library, an FDB application uses, to access the database. It exposes the transaction concept to applications. Client in FDB is a fat client that does multiple complex operations: (1) It calculates read and write conflict ranges for transactions; (2) it batches a transaction’s operations and send them all together at commit for better throughput; (3) it automatically retries failed transactions.

File: bench/bench_test.go (hunk 7 above)
Comment (gammazero) [reply bench/bench_test.go:171]: Nevermind, must have read the diff backward.

File: fdb/fdb.go (hunk 8 above)
Comment (gammazero) [reply fdb/fdb.go:79]: OK, proceed.
